compiler/fir/checkers/src/org/jetbrains/kotlin/fir/analysis/diagnostics/coneDiagnosticToFirDiagnostic.kt

```diff
@@ -88,7 +89,8 @@ private fun mapUnsafeCallError(
     if (rootCause !is InapplicableWrongReceiver) return null
     val actualType = rootCause.actualType ?: return null
     val expectedType = rootCause.expectedType
-    if (actualType.isNullable && (expectedType == null || expectedType.isEffectivelyNotNull())) {
+    if (actualType.canBeNull && (expectedType == null || expectedType.isEffectivelyNotNull())) {
```

Comment:
Why do we need this separate `isEffectivelyNotNull` method? It seems `!expectedType.canBeNull` would just work.

In addition, it seems to me that the current `canBeNull` implementation has a few bugs:

1. in `is ConeTypeParameterType -> this.lookupTag.typeParameterSymbol.fir.bounds.any { it.canBeNull }`, checking the bounds should be done with `all` instead of `any`.
2. we should add a case for `ConeTypeVariableType`,  someting like `is ConeTypeVariableType -> ((this.lookupTag as? ConeTypeVariableTypeConstructor)?.originalTypeParameter as? ConeTypeParameterLookupTag)?.typeParameterSymbol?.fir?.bounds?.any { it.canBeNull } ?: isNullable`.

With 2, the problem that Jinseong mentioned (https://github.com/JetBrains/kotlin/pull/4266#discussion_r605833315) in [52.fir.kt](https://github.com/JetBrains/kotlin/pull/4266/files/eca7638e2a1b56b9ca2d52723ae8a39653a8f866#diff-c32361960945293060dd5f64f5ede993b329b4f247082e734ffaa6bc56fdf7d4) is addressed: that error should not be an `UNSAFE_CALL`, but instead an `INAPPLICABLE_CANDIDATE`. Note, however, it seems this particular error shouldn't have been reported in the first place. But similar issues could make the current logic report wrong diagnostics. For example, with `UPPER_BOUND_VIOLATED` that you have mentioned. There is no need to differentiate this particular problem (maybe more in future?) if the null check was done correctly.

Reply:
I eliminated `isEffectivelyNotNull`. I changed to use `all` instead of `any` for `ConeTypeParameterType` and added a test (see 2nd commit), and found that the same test is broken in FE1.0 (both IR and non-IR backends): https://youtrack.jetbrains.com/issue/KT-45903.

And as you pointed out, 52.fir.kt is no longer changed. Thanks for the suggested fix!